### PR TITLE
[Ubuntu] Install aws-sam-cli without brew

### PR DESCRIPTION
**Pull request by miketimofeev**
# Description
Aws-sam-cli is supposed to be installed via homebrew yet that approach leads to install separate python3, which overrides system one and causes issues for some customers. The workaround was implemented here https://github.com/actions/virtual-environments/pull/789 but it turned out that brew functionality is partly broken after the changes.
The solution is to get-rid of brew aws-sam cli installation and use `setup.py` script that comes along with aws-sam sources. After these changes, another PR will be made to move homebrew to the beginning of the PATH as it was initially.

Changes:
- Aws-sam-cli installation switched to build from sources via setup.py script
- Installation script moved to the end of the template in order to run after toolset installation

#### Related issue:
https://github.com/actions/virtual-environments/issues/1136

## Check list
- [x] Related issue / work item is attached
- [x] Tests are written (if applicable)
- [x] Documentation is updated (if applicable)
- [ ] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/linux/scripts/installers/aws-sam-cli.sh [hunk 1]
```diff
@@ -2,15 +2,31 @@
...
 # Source the helpers for use with the script
 source $HELPER_SCRIPTS/document.sh
 
-# Install aws sam cli
-brew tap aws/tap
-brew install aws-sam-cli
+# Download latest aws sam cli sources
+TarballUrl=$(curl -s https://api.github.com/repos/awslabs/aws-sam-cli/releases/latest | jq -r '.tarball_url')
+TarballPath="/tmp/aws-sam-cli.tar.gz"
+wget $TarballUrl -O $TarballPath
```
Comment (vsafonkin) [inline images/linux/scripts/installers/aws-sam-cli.sh:14]: Does it make sense to use `download_with_retries()` helper?

Comment (miketimofeev) [reply images/linux/scripts/installers/aws-sam-cli.sh:14]: I've tried but it looks like curl can't download from such links with redirection or maybe some flag is supposed to be specified

File: images/linux/ubuntu1604.json [hunk 2]
```diff
@@ -288,6 +276,18 @@
...
         },
+        {
+            "type": "shell",
+            "scripts": [
+                "{{template_dir}}/scripts/installers/aws-sam-cli.sh"
+            ],
+            "environment_vars": [
+                "METADATA_FILE={{user `metadata_file`}}",
+                "HELPER_SCRIPTS={{user `helper_script_folder`}}",
+                "DEBIAN_FRONTEND=noninteractive"
```
Comment (al-cheb) [inline images/linux/ubuntu1604.json:287]: Do we need it?

File: images/linux/scripts/installers/aws-sam-cli.sh [hunk 3]
```diff
@@ -2,15 +2,31 @@
...
+tar -xzvf $TarballPath -C /tmp
+SourcesDir=$(echo /tmp/awslabs-aws-sam-cli*)
+
+# Use python 3.7 from toolcache to install aws sam, setuptools package required for the installation
+Python3Dir=$(echo ${AGENT_TOOLSDIRECTORY}/Python/3.7*/x64)
+Python3BinDir=$(echo ${Python3Dir}/bin)
```
Comment (al-cheb) [inline images/linux/scripts/installers/aws-sam-cli.sh:20]: nipick: "${Python3Dir}/bin"

File: images/linux/scripts/installers/aws-sam-cli.sh [hunk 4]
```diff
@@ -2,15 +2,31 @@
...
 
 # Source the helpers for use with the script
 source $HELPER_SCRIPTS/document.sh
 
-# Install aws sam cli
-brew tap aws/tap
-brew install aws-sam-cli
+# Download latest aws sam cli sources
+TarballUrl=$(curl -s https://api.github.com/repos/awslabs/aws-sam-cli/releases/latest | jq -r '.tarball_url')
+TarballPath="/tmp/aws-sam-cli.tar.gz"
```
Comment (AlenaSviridenko) [inline images/linux/scripts/installers/aws-sam-cli.sh:13]: I think our bash variables naming convention is either ALL_UPPER_CASE or all_lower_case symbols with underscores.

File: images/linux/scripts/installers/aws-sam-cli.sh (hunk 3 above)
Comment (miketimofeev) [reply images/linux/scripts/installers/aws-sam-cli.sh:20]: Thanks!

File: images/linux/ubuntu1604.json (hunk 2 above)
Comment (miketimofeev) [reply images/linux/ubuntu1604.json]: Nope, forgot to remove it, thanks!

File: images/linux/scripts/installers/aws-sam-cli.sh (hunk 4 above)
Comment (miketimofeev) [reply images/linux/scripts/installers/aws-sam-cli.sh:13]: Judging by the other scripts we don't have any convention 😀
I've seen both your options and Julia.sh contains the same name types
image

File: images/linux/scripts/installers/aws-sam-cli.sh [hunk 5]
```diff
@@ -2,15 +2,31 @@
...
+wget $TarballUrl -O $TarballPath
+tar -xzvf $TarballPath -C /tmp
+SourcesDir=$(echo /tmp/awslabs-aws-sam-cli*)
```
Comment (al-cheb) [inline images/linux/scripts/installers/aws-sam-cli.sh:16]: I think we can rid of SourcesDir variable and move cd from 23 line:
`cd /tmp/awslabs-aws-sam-cli*`

File: images/linux/scripts/installers/aws-sam-cli.sh [hunk 6]
```diff
@@ -2,15 +2,31 @@
...
+# Use python 3.7 from toolcache to install aws sam, setuptools package required for the installation
+Python3Dir=$(echo ${AGENT_TOOLSDIRECTORY}/Python/3.7*/x64)
+Python3BinDir=$(echo ${Python3Dir}/bin)
+export PATH="$Python3Dir:$Python3BinDir:$PATH"
+python3 -m pip install setuptools
+cd $SourcesDir
+python3 setup.py install
+sudo ln -sf ${Python3BinDir}/sam /usr/local/bin/sam
+
+# Cleanup downloaded files
```
Comment (al-cheb) [inline images/linux/scripts/installers/aws-sam-cli.sh:27]: nitpick: we can remove this section and leave it the `cleanup.sh` script

File: images/linux/scripts/installers/aws-sam-cli.sh (hunk 4 above)
Comment (AlenaSviridenko) [reply images/linux/scripts/installers/aws-sam-cli.sh:13]: ok than, disregard my comment

File: images/linux/scripts/installers/aws-sam-cli.sh (hunk 6 above)
Comment (miketimofeev) [reply images/linux/scripts/installers/aws-sam-cli.sh:26]: Cool, thanks!

File: images/linux/scripts/installers/aws-sam-cli.sh [hunk 7]
```diff
@@ -2,15 +2,26 @@
 ################################################################################
 ##  File:  aws-sam-cli.sh
 ##  Desc:  Installs AWS SAM CLI
-##         Must be run as non-root user after homebrew and clang
+##         Requires Pyton >=3.6, must be run as non-root user after toolset installation
```
Comment (maxim-lobanov) [inline images/linux/scripts/installers/aws-sam-cli.sh:5]: typo: `Pyton`

File: images/linux/scripts/installers/aws-sam-cli.sh (hunk 1 above)
Comment (miketimofeev) [reply images/linux/scripts/installers/aws-sam-cli.sh:14]: yeah, -L flag should be specified, which is not implemented in the download_with_retries function

**Pull request merged**